src/client/datascience/jupyter/kernels/kernelExecution.ts

```diff
@@ -33,26 +34,20 @@ const vscodeNotebookEnums = require('vscode') as typeof import('vscode-proposed'
  * Else the `Kernel` class gets very big.
  */
 export class KernelExecution implements IDisposable {
-    private readonly cellExecutions = new WeakMap<NotebookCell, CellExecution>();
-
-    private readonly documentExecutions = new WeakMap<NotebookDocument, CancellationTokenSource>();
-    private readonly stackOfCellsToExecuteByDocument = new WeakMap<NotebookDocument, CellExecution[]>();
-
+    private readonly documentExecutions = new WeakMap<NotebookDocument, CellExecutionQueue>();
```

Comment:
This is where the refactor helps. Some are in the new cellExecutionQueue class, but previously there were just too many tracking queues and maps being juggled here. cellExecutions, documentExecutions, stackOfCellsToExecuteByDocument, kernelValidated, pendingExecution, pendingCellExecution. This class is much improved now, even if we still have to deal with some complexity in the cellExecutionQueue.